feat(api): Cancel in flight api request

kbuechl opened the pull request:
*Issue #, if available:*
Fixes #989 

*Description of changes:*
Added cancel token options to rest api as well as graphql. Aliased axios' `isCancel` as well as `CancelToken` to keep the interface familiar.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

kbuechl commented:
I am feeling like I could add more to unit tests but am not sure how to further improve them without mocking the logic behind cancel tokens that axios utilizes currently, and that feels like overkill.  I am open to suggestions on the unit test coverage.

pravgupt commented on packages/api/src/API.ts:
```diff
@@ -328,7 +345,7 @@ export default class APIClass {
      * @returns {Promise<GraphQLResult> | Observable<object>}
      */
-    graphql({ query: paramQuery, variables = {}, authMode }: GraphQLOptions) {
+    graphql({ query: paramQuery, variables = {}, authMode, cancelToken }: GraphQLOptions) {
```
I believe it's better to update the graphql API similar to the existing REST API schema, i.e. by accepting an init object for _extra_ parameters related to the request.

kbuechl replied:
Should this be a part of the GraphQLOptions object, or would you rather it was a separate parameter all together?

kbuechl replied:
Additionally, should i use the type from axios for the cancel token or leave it as any? based on other the other comments about leaving axios specific logic out of the library, I am guessing the later but want to double check.

kbuechl replied:
@pravgupt Do you want to double check this is how you want it?

pravgupt commented on packages/api/src/API.ts:
```diff
@@ -50,6 +52,21 @@ export default class APIClass {
+     * Create a new cancel token
+     * @return {CancelTokenStatic} - Axios cancel token
+     */
+    readonly CancelToken: CancelTokenStatic = axios.CancelToken;
```
Are you using this token anywhere in the library? I believe tokens are generated and passed in by the client application.

kbuechl replied:
So I am not using it anywhere in the library, but wanted to expose it so that axios would not need to be imported into the client application in order to use this. I should have added the example usage code i put in the issue, but this would allow the client application to initialize a new token by just using it from the API instance.

```javascript 
const { API } = require('aws-amplify');
const cancelTokenSource = API.CancelToken.source();
const init = {
   cancelToken: cancelTokenSource.token,
   //...
};
```

pravgupt commented on packages/api/src/API.ts:
```diff
@@ -50,6 +52,21 @@ export default class APIClass {
+    * @param {any} value - the exception thrown from an api call
+    * @return {boolean} - A boolean indicating if the exception was a cancellation
+    */
+    isCancel(value: any): boolean {
```
Same as above, if we are not using this in the library and not exporting, see if we can get rid of it.

kbuechl replied:
I was thinking the same as above for this one, just didn't want to force anyone to import axios. When an API call is canceled, it will result in an error that can be checked with this function. What are your thoughts?

pravgupt replied:
I'd prefer to keep the library clean of axios specific functionality and in this instance let the clients pass in the cancel token to keep library more flexible for future updates.

kbuechl replied:
This would require the clients to install axios if they want to utilize cancellation tokens (and subsequently checking errors thrown by cancellations), is this alright?

I am concerned that this will require the client to  understand that axios is used under the covers and could lead to additional packages in consuming applications that would be unused if the api core logic moved off of amplify. Are there any other options that would prevent consumers to install axios, but also protect the library from being too locked into axios? Is there another way to export it that you would be comfortable with?

kbuechl replied:
This would change the consumers implementation to be similar to this

```javascript
const { API } = require('aws-amplify');
const { CancelToken, isCancel } = require('axios');
var cancelTokenSource = CancelToken.source();
const init = {
	cancelToken: cancelTokenSource.token,
	//...
};
try {
	let response = await API.get(API_NAME, PATH, init);
}
catch (err) {
	if (isCancel(err)) {
		//request cancelled
	}
	//other exception
}
```

Then when you wanted to cancel
```javascript
cancelTokenSource.cancel("Tired of waiting");
```

pravgupt replied:
I think I'm ok with this having clients who need cancel functionality to import and manage cancel tokens. Why I am a little hesitant to make it Amplify's API interface is that the axios cancel token API is [based on the](https://github.com/axios/axios#cancellation) withdrawn cancelable promises proposal.

kbuechl replied:
That makes sense, thanks for giving me a bit more context into your thoughts!

pravgupt commented on packages/api/package.json:
```diff
@@ -27,6 +27,7 @@
   },
   "homepage": "https://github.com/aws/aws-amplify#readme",
   "devDependencies": {
+    "@aws-amplify/pubsub": "^1.0.31",
```
Is this needed?

kbuechl replied:
I was unable to run the unit tests inside of /api without this, it was referenced in the graphql tests.

kbuechl replied:
I am going to remove this and see if this is just something I did incorrectly, but I think this should be a part of the dependencies instead of devDependencies since core, cache, and auth are already listed in the dependencies

kbuechl replied:
I removed this and the circlci unit tests passed just fine. I will look into why it isnt working locally and keep this out.

kbuechl commented:
I am out of town this weekend and this is all on mobile so sorry about the poor formatting. I will look at the graphql suggestion when I get back this week.

houmark commented:
What happened on this one? Do you need help to bring this to approval? I'd really like to see this merged.

kbuechl commented:
@houmark I am still awaiting a reply on a few questions I had. I have some questions about removing the aliased axios parts and what the ramifications of this would be. 

@pravgupt Have you had a chance to review my questions in regards to this?

kbuechl commented:
@pravgupt I think this is ready to go. Let me know if you see anything else

jpbourgeon commented:
I follow @houmark  on this one

This feature would open the way for interesting optimisations in many UX cases like on-the-fly searches, autocompletion, infinite-scrolling...

Is there a way we can help it land into a close release ?

lissau commented:
I'm really looking forward to this feature!

Any updates here?

ericclemmons commented on packages/api/src/API.ts:
```diff
@@ -374,14 +374,14 @@ export default class APIClass {
         };
 
-        const init = {
+        Object.assign(init, {
```
Since the `init` object is potentially coming from a user, `Object.assign` will may be mutating a reference to their object.

```js
const init = { cancelToken: 1 }
// undefined
Object.assign(init, { body: "isAWonderland" })
// Object { cancelToken: 1, body: "isAWonderland" }
init
// Object { cancelToken: 1, body: "isAWonderland" }
```

This can be rewritten without mutation by replacing https://github.com/aws-amplify/amplify-js/blob/a4e5fbe8feed03c38a876118518d373810831903/packages/api/src/API.ts#L399

with:

```js
response = await this._api.post(
  endpoint,
  Object.assign({}, init, { headers, body, signerServiceInfo })
);
```

kbuechl replied:
Took care of this. This was an oversight on my part, good catch thank you!

ericclemmons commented on packages/api/src/RestClient.ts:
```diff
@@ -101,13 +102,16 @@ export class RestClient {
             params.timeout = initParams.timeout;
         }
+        if (initParams.cancelToken) {
+            params.cancelToken = initParams.cancelToken;
```
Can you link/share docs, code, or URL for how this is handled downstream?

This PR passes a user's `cancelToken` through the API, but I'm curious where the `AbortController` is being leveraged.

kbuechl replied:
@ericclemmons Here is the documentation from axios about [cancellations](https://github.com/axios/axios/blob/master/README.md#cancellation). I am not quite sure I understand what you mean about the `AbortController` being leveraged, can you explain a bit more? It would be up to the consumer who provides the token to utilize the cancellation token in their implementation. I have a few examples in the related issue and I believe I added them to previous responses in here as well but let me know if you want me to post them again.

houmark commented:
@jpbourgeon @lissau @ericclemmons This one got stuck when these changes were requested:
https://github.com/aws-amplify/amplify-js/pull/3769#pullrequestreview-285280161

I'm anxiously awaiting this to land also. We have had many race situations in our UI without any easy way to work around it, especially when querying large data, returning more than 20-30 results or searching all data. Our UI is made so that the user can click and check off different filters and a filter change results in an API request, but it does not finish before the next click and sometimes as you are filtering for less data the second filter will finish before the first one, which then lists the wrong results for the current filter :/

@undefobj hoping you can get this one back on track with @manueliglesias as it seems there's quite a lot of devs out there that would use this.

Amplifiyer commented:
@kbuechl, I apologize for the delay, we had an internal API review today for this feature and decided to bring the axios specific implementation in the library instead of making customers rely on it or leaking any axios implementation to the customers. Below is the proposal we agreed on

Amplify library will expose an API to cancel any specific request. Library will keep a mapping of every request's promise with an axios token (generated by the amplify library) in a `WeakMap` . When customers want to cancel a request, they will call this new API and pass the promise object they got from the actual API call.

```
const responsePromise = API.graphQL({});
.
.
.
// Customer wants to cancel the request
API.cancel(responsePromise);
```

Amplify library will also expose another API called `isCancel(exception)` that will be a wrapper to `axios.isCancel()` to let customers determine if an exception/throwable is due to request being cancelled.

This approach allows customers to not depend on axios, and hence allows Amplify library to seamlessly switch to a different rest provider (such as fetch) in the future without impacting customers.

@kbuechl, please let us know if you'll be able to implement this proposal. We will prioritize review of your PR.

kbuechl commented:
Just wanted to follow up on this, Thank you for talking about it more in depth, I think this is a really great direction to go. I can definitely take care of this, sorry for the delay I have been traveling quite a bit recently and am headed to reinvent next week. Should be able to start this next week or the week after.

Amplifiyer commented:
> Just wanted to follow up on this, Thank you for talking about it more in depth, I think this is a really great direction to go. I can definitely take care of this, sorry for the delay I have been traveling quite a bit recently and am headed to reinvent next week. Should be able to start this next week or the week after.
> 
> @Amplifiyer Would it be best for me to update this branch and this PR or should I cancel this and start a new one?

@kbuechl Updating this branch would be fine though there are few merge conflicts. If you would rather start fresh, go for it. Thanks again for doing this!

kbuechl commented:
@Amplifiyer I wanted to circle back on this. I made the code changes but I cannot figure out how to unit test the cancel token portion of axios properly. I do not know if the limitation I am running into is because of the way axios setup the `canceltoken` object itself, or a limitation of mocking default exports next to named exports in jest, but I am honestly out of ideas to solve this and do not want this to continue waiting on me as I think it is really useful functionality. What would be the best way to continue at this point?

Amplifiyer commented:
> @Amplifiyer I wanted to circle back on this. I made the code changes but I cannot figure out how to unit test the cancel token portion of axios properly. I do not know if the limitation I am running into is because of the way axios setup the `canceltoken` object itself, or a limitation of mocking default exports next to named exports in jest, but I am honestly out of ideas to solve this and do not want this to continue waiting on me as I think it is really useful functionality. What would be the best way to continue at this point?

Thanks @kbuechl for your effort. Can you push your changes(including the current state of the unit tests) to a fork where I can pull from and look at more closely?

kbuechl commented:
@Amplifiyer I added a few examples of the various things I had tried and what I was having some luck with and others I was not. Even when I am able to get the tests to pass, I cannot get the mock for source to register that it is being called even though it is providing the correct response. It is also worth noting I have not looked much at the graphql portion of this (based on where I added the new code it will be implemented for both the restclient and graphql, but I have not attempted to test yet).

I really don't want to throw in the towel on this and would be happy to finish this if you can help me out figure out what I have been doing wrong with the tests just let me know.

Let me know if there are any problems seeing my changes [here](https://github.com/kbuechl/amplify-js/tree/cancelApiRequests)

Amplifiyer commented:
@kbuechl Thanks, I'll look into it this week.

houmark commented:
Hey guys,

Did this one lose traction again? Really hoping to see this one make it soon.

Amplifiyer commented:
> Hey guys,
> 
> Did this one lose traction again? Really hoping to see this one make it soon.

@houmark , I've been looking into it this week but it's proving difficult as @kbuechl mentioned to mock axios cancelToken interface. I would need to deep dive into axios source code for some clues. Will update in couple days if I make some more progress here.

Amplifiyer commented:
@kbuechl Please check https://github.com/kbuechl/amplify-js/commit/340cfce04b0cb97c031de24810c9bb15a28ee63e, I've been able to fix the existing tests using jest.spyon. Let me know if you face any issues with it.

houmark commented:
@kbuechl @Amplifiyer what's holding this one back currently?

minaairsupport commented:
could you please push releasing this PR we need it badly

Amplifiyer commented:
@kbuechl, please let us know if you were able to use  https://github.com/kbuechl/amplify-js/commit/340cfce04b0cb97c031de24810c9bb15a28ee63e to fix the unit tests and are close to re-publishing the PR.

kbuechl commented:
Hey all, sorry my time has been the only thing holding this up. I took a look at the code but have not had a chance to work on this recently. I am planning to check it out this evening will update everyone after I get I to it. 

@amplifiyer @houmark

houmark commented:
Thank you @kbuechl, really looking forward to it!

kbuechl commented:
@Amplifiyer Thank you for your code. the changes you suggested did the trick and I was able to continue. One last sticking point I am working through is confirming the request cancel is behaving properly. I am finding that when the test runs the request is added to the weakmap fine but when the following line to cancel the request is called the weakmap is empty all of the sudden, meaning that my mock for cancel is never called. I am going to look into it tomorrow and see if I can figure out why it keeps clearing.

sammartinez commented:
@kbuechl This is looking pretty great! Is there any way you can resolve the conflicts within this Pull Request? We would like your implementation if possible for this feature. Let us know if you do not have time and we can take over this pull request on your behalf. Thanks ahead of time!

kbuechl commented:
Hey everyone, Sorry I haven't been able to prioritize this more recently. The only hang up is the one failing unit test, as it appears that the weakmap is clearing out by the time the cancel code is run. Things have gotten busy for me recently and this has just not stayed high on my priority list. If anyone is able to figure out the issue with the unit tests I would be more than happy to take some help on this so we can get it wrapped up. @sammartinez @houmark @Amplifiyer

sammartinez commented:
> @kbuechl, please let us know if you were able to use [kbuechl@340cfce](https://github.com/kbuechl/amplify-js/commit/340cfce04b0cb97c031de24810c9bb15a28ee63e) to fix the unit tests and are close to re-publishing the PR.

@kbuechl This looks like the fix for the unit tests were here. Do you want to double check off of this commit?

Amplifiyer commented:
> The only hang up is the one failing unit test, as it appears that the weakmap is clearing out by the time the cancel code is run.

@kbuechl, will you be able to push your current changes in your branch and we can look at it?

kbuechl commented:
@Amplifiyer I have my most recent code pushed. You will see when you run the tests the [happy case](https://github.com/kbuechl/amplify-js/blob/60053865a10ce220158e8f7d63b1c26fb7c8516e/packages/api/__tests__/RestClient-unit-test.ts#L478) for the token cancellation is what is failing.

@sammartinez No that did fix the initial mock problems we had, What I have found now is that this one test (see above) is failing, when debugging it the initial call shows that it is loading into the weakmap properly but when `cancel` is called the weakmap is empty all of the sudden.

Amplifiyer commented:
@kbuechl, I tried to fix the unit test failing and ended up realizing that the failing unit test has revealed a novel bug in the implementation. The issue with implementation is that the axios promise that we are saving in the weakmap is not the same that client are getting from the rest client due to another promise being wrapped in async function and due to promise's auto-unwrap, there is no way to get the nested promise.

Not sure if what I said above is making sense, but below is a small code snippet to explain

```js
var aConstantPromise = Promise.resolve("data"); // Mimics axios promise which we save in weakmap
var anAsyncFunction = async function() {    // mimics ajax function of RestClient
    return aConstantPromise;
}
var returnedPromise = anAsyncFunction(); // Promise that clients receive
console.log(aConstantPromise === returnedPromise); // false -> Not the same
returnedPromise.then(console.log); // prints data -> No way to get the original promise, i.e. aConstantPromise
```

I've hit a blocker here to come up with a clean solution. The only thing I can think of for this design solution is to make ajax() method non-async and convert this._custom_header() to then.

houmark commented:
This feels like it's so close — really hoping to see this make the finish line soon.

ngault commented:
@Amplifyer   A lot of folks have been desperately watching this issue and its parent #989 for 2 years now.  No doubt the AWS maintainers are busy, but it's hard to accept that such a fundamental, critical capability as aborting is still missing from amplify-api.  

This is a capability that is assumed by any framework (React, Vue, etc) and without which performant modern web apps can't be written.  

If you're not able to bring this home, maybe it's time to ask for help?

houmark commented:
I agree completely, and this and other issues like no possibility of counting results, general instability and breaking changes without a clear change log both here and in the CLI, is why we are close to making the decision to commence migration to a new platform as soon as we can allocate time.

That brings the lifetime of Amplify for us down ~10 months, which is maybe the shortest living platform I have made in my 22-year career. It's sad, and I personally feel bad for deciding on this platform, because we now have 35k lines of code to migrate to another platform and it's sad that the largest cloud provider in the world cannot pull this off. You should think they should be able to allocate the necessary resources if they wanted to, but it seems they may already have given up on Amplify internally and have gone into some kind of maintenance mode.

sammartinez commented:
All,

I wanted to comment and also state out apologizes on behalf of Amplify for this feature request and not being able to close this on a timely matter. I appreciate everyone’s patience on this and working with us in order to get this out to the world. The one thing that @Amplifiyer callout above is with us resolving the unit tests, this opened up another can of worms of issues and breaking changes to our library. With that being stating, we are brainstorming internally on this for the right solution and finding an owner on our team in order to get this out sooner rather than later. That being stated, we appreciate the feedback on this and I will be following up next week on this to give more transparency on if we have an outcome from brainstorming and a owner for this feature. 

Thanks again for the feedback on this !

ngault commented:
@sammartinez , thanks for getting involved.  You express appreciation for feedback, so, as one of many devs with a huge amount riding on improvement to Amplify, here is some more:

1) You'll need to do what the Amplify dev team historically has not: immediately propose a timeline with specifics, and then deliver on it.  Otherwise, nobody here will take you seriously.  The Amplify team is facing a much broader crisis in credibility.  This complaint shouldn't be viewed as isolated.

2) aws-amplify needs more resources from AWS -- both more developers and more senior developers.  The large AWS customers using Amplify understand that this project is in deep distress, and they all attribute that lack of funding.  For so many of us, this issue is but one of a large number that are causing us needless expense and delays in our own offerings.   I don't know @houmark above, but he is like a lot of other Amplify users I know that are in the midst of evaluating alternative platforms.

I would tell your management this: Amplify's struggles are threatening AWS other businesses.  To the extent you're the free "on-ramp" to the rest of the AWS cloud platform, you're now driving millions of dollars of revenue to AWS' competitors.

Amplifiyer commented:
@kbuechl, I had create another PR because resolving merge conflicts is a big pain due to api category split that happened a month ago.

I've been able to change the approach slightly to get around the promise wrapping issue and fixed the unit tests. Please review https://github.com/aws-amplify/amplify-js/pull/5753

sammartinez commented:
All,

After discussions internally, we were able to come to an agreement and the above pull request has been merged into our unstable branch. We will look to do a release sometime this week, just wanted to provide and update and that we have closed this PR in favor of the above.

houmark commented:
Great. Finally!!! A tiny bit of documentation would be nice in order to
kick the wheels on this in the testing branch of our app.

Feedback will come promptly.

Thanks again for finally pulling this one through.

- Lars

On 13 May 2020, at 14.16, Sam Martinez <notifications@github.com> wrote:

﻿

All,

After discussions internally, we were able to come to an agreement and the
above pull request has been merged into our unstable branch. We will look
to do a release sometime this week, just wanted to provide and update and
that we have closed this PR in favor of the above.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/aws-amplify/amplify-js/pull/3769#issuecomment-628192350>,
or unsubscribe
<https://github.com/notifications/unsubscribe-auth/AADSDATGLHA2NZMGOYEVMZ3RRLWZDANCNFSM4IIDSK7A>
.

kbuechl commented:
> @kbuechl, I had to create another PR because resolving merge conflicts is a big pain due to api category split that happened a month ago.
> 
> I've been able to change the approach slightly to get around the promise wrapping issue and fixed the unit tests. Please review #5753

@Amplifiyer Thanks for seeing this to the end!
